src/signTx.c

```diff
@@ -1664,6 +1692,166 @@ static void signTx_handleScriptDataHashAPDU(uint8_t p2, uint8_t* wireDataBuffer,
+{
+	TRACE_STACK_USAGE();
+	{
+		// sanity checks
+		CHECK_STAGE(SIGN_STAGE_BODY_COLLATERALS);
+		ASSERT(BODY_CTX->currentCollateral < ctx->numCollaterals);
+
+		VALIDATE(p2 == P2_UNUSED, ERR_INVALID_REQUEST_PARAMETERS);
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+
+	BODY_CTX->stageData.collateral = extractTransactionInput(wireDataBuffer, wireDataSize);
+
+	{
+		// add to tx
+		TRACE("Adding collateral to tx hash");
+		txHashBuilder_addCollateral(
+		        &BODY_CTX->txHashBuilder,
+		        BODY_CTX->stageData.collateral.txHashBuffer, SIZEOF(BODY_CTX->stageData.collateral.txHashBuffer),
+		        BODY_CTX->stageData.collateral.parsedIndex
+		);
+	}
+
+	security_policy_t policy = policyForSignTxCollaterals(ctx->commonTxData.txSigningMode);
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+
+	{
+		// select UI steps
+		switch (policy) {
+#	define  CASE(POLICY, UI_STEP) case POLICY: {ctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_COLLATERAL_STEP_DISPLAY);
+			CASE(POLICY_ALLOW_WITHOUT_PROMPT, HANDLE_COLLATERAL_STEP_RESPOND);
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+	}
+	signTx_handleCollateral_ui_runStep();
+}
+
+// ========================= REQUIRED SIGNERS ===========================
+
+enum {
+	HANDLE_REQUIRED_SIGNERS_STEP_DISPLAY = 1400,
+	HANDLE_REQUIRED_SIGNERS_STEP_RESPOND,
+	HANDLE_REQUIRED_SIGNERS_STEP_INVALID,
+};
+
+static void signTx_handleRequiredSigner_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	ui_callback_fn_t* this_fn = signTx_handleRequiredSigner_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step, this_fn);
+
+	UI_STEP(HANDLE_REQUIRED_SIGNERS_STEP_DISPLAY) {
+		ui_displayHexBufferScreen("Required signer", BODY_CTX->stageData.requiredSigner, SIZEOF(BODY_CTX->stageData.requiredSigner), this_fn);
+	}
+
+	UI_STEP(HANDLE_REQUIRED_SIGNERS_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+
+		// Advance stage to the next input
+		ASSERT(BODY_CTX->currentRequiredSigners < ctx->numRequiredSigners);
+		BODY_CTX->currentRequiredSigners++;
+
+		if (BODY_CTX->currentRequiredSigners == ctx->numRequiredSigners) {
+			advanceStage();
+		}
+	}
+	UI_STEP_END(HANDLE_REQUIRED_SIGNERS_STEP_INVALID);
+}
+
+__noinline_due_to_stack__
+static void signTx_handleRequiredSignerAPDU(uint8_t p2, uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	TRACE_STACK_USAGE();
+	{
+		// sanity checks
+		CHECK_STAGE(SIGN_STAGE_BODY_REQUIRED_SIGNERS);
+		ASSERT(BODY_CTX->currentRequiredSigners < ctx->numRequiredSigners);
+
+		VALIDATE(p2 == P2_UNUSED, ERR_INVALID_REQUEST_PARAMETERS);
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+
+	{
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+		STATIC_ASSERT(SIZEOF(BODY_CTX->stageData.requiredSigner) == VKEY_LENGTH, "wrong vkey length");
+		view_copyWireToBuffer(BODY_CTX->stageData.requiredSigner, &view, VKEY_LENGTH);
```

Comment:
I think it makes sense to allow derivation paths, too. HW wallet users probably think in terms of paths, not actual keys which are much harder to remember or compare.